src/build/dlopen.c

```diff
@@ -0,0 +1,62 @@
+/**
+ * @file dlopen.c
+ * @brief Proof of concept for calling dlopen() from GNU make.
+ *
+ * @author James Jones (jejones@networkradius.com)
+ * @copyright 2020 Network RADIUS SARL (legal@networkradius.com)
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <dlfcn.h>
+#include <gnumake.h>
+
+#ifdef __APPLE__
+#  define DL_EXTENSION ".dylib"
+#elif defined (WIN32)
+#  define DL_EXTENSION ".dll"
+#else
+#  define DL_EXTENSION ".so"
+#endif
+
+char	*mk_dlopen(char const *nm, unsigned int argc, char **argv);
+int	dlopen_gmk_setup(void);
+
+int plugin_is_GPL_compatible;
+
+/** GNU make-callable layer around dlopen().
+ * 
+ * @return NULL on failure; otherwise, a string. 
+ * 
+ * @note argv[0] should point to the name of the file to open, minus the extension.
+ * 
+ */
+char * mk_dlopen(UNUSED char const *nm, UNUSED unsigned int argc, char **argv)
+{
+	void	*handle;
+	char	*extended_file_name;
+	char	*result;
+	
+	extended_file_name = gmk_alloc(1 + strlen(DL_EXTENSION) + strlen(argv[0]));
+	sprintf(extended_file_name, "%s%s", argv[0], DL_EXTENSION);
+
+	handle = dlopen(extended_file_name, RTLD_LAZY);
+
+	if (!handle) return NULL;
+	
+	result = gmk_alloc(2);
+	strcpy(result, "1");
```

Comment:
can you please `dlclose` the handle.

Reply:
Returning "1" isn't overly useful.  It'd be better if we could get the path name to the dynamic library.

On Solaris, FreeBSD and Linux this information is available from `dlinfo()`, querying `RTLD_DI_ORIGIN`.

On macos you can iterate over the loaded images, and find the one that matches the pointer returned by `dlopen()`. `_dyld_get_image_name()` will apparently return the absolute path.

Example code is available here: https://stackoverflow.com/questions/20481058/find-pathname-from-dlopen-handle-on-osx man page is available here: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dyld.3.html

Reply:
I will dlclose() the handle and return the path name.

Reply:
Shall I not concern myself with Windows for the time being?

Reply:
@jejones3141 We don't run on Windows.  :(  There was some attempt to port it to Windows ~10 years ago.  But most of that has been stripped out.

Reply:
Yeah, honestly you could strip the windows case out of the extension macro.